packages/core/src/components/forms/asyncControllableInput.tsx

```diff
@@ -134,17 +142,21 @@ export class AsyncControllableInput extends React.PureComponent<
     }
 
     private handleCompositionStart = (e: React.CompositionEvent<HTMLInputElement>) => {
-        this.setState({
-            isComposing: true,
-            // Make sure that localValue matches externalValue, in case externalValue
-            // has changed since the last onChange event.
-            nextValue: this.state.value,
-        });
+        this.cancelPendingCompositionEnd?.();
+        this.setState({ isComposing: true });
         this.props.onCompositionStart?.(e);
     };
 
     private handleCompositionEnd = (e: React.CompositionEvent<HTMLInputElement>) => {
-        this.setState({ isComposing: false });
+        // In some non-latin languages, a keystroke can end a composition event and immediately afterwards start another.
+        // This can lead to unexpected characters showing up in the text input. In order to circumvent this problem, we
+        // use a timeout which creates a delay which merges the two composition events, creating a more natural and predictable UX.
+        // `this.state.nextValue` will become "locked" (it cannot be overwritten by the `value` prop) until a delay (10ms) has
+        // passed without a new composition event starting.
+        this.cancelPendingCompositionEnd = this.setTimeout(
```

Comment:
@Athenodoros we have a mechanism in `AbstractPureComponent2` to cancel outstanding timeouts / intervals when components unmount, so I've gone ahead and used that here instead of the lower-level `window.setTimeout`